fix(usb/hid): HID Host Driver v.1.0.1

roma-jam opened the pull request:
# Checklist

- [X] Component contains License
- [X] Component contains README.md
- [X] Component contains idf_component.yml file with `url` field defined
- [X] Component was added to [upload job](https://github.com/espressif/idf-extra-components/blob/master/.github/workflows/upload_component.yml#L18)
- [X] Component was added to [build job](https://github.com/espressif/idf-extra-components/blob/master/test_app/CMakeLists.txt#L8)
- [X] _Optional:_ Component contains unit tests
- [X] CI passing

# Changelog
- Fixed a bug where configuring the driver with `create_background_task = false` did not properly initialize the driver. This lead to `the hid_host_uninstall()` hang-up.
- Fixed bug where `hid_host_uninstall()` would cause a crash during the call while USB device has not been removed.
- Added `hid_host_get_device_info()` to get the basic information of a connected USB HID device.

## Behavior description while create_background_task=false
### Brief description
Fix `hid_host_uninstall()` hanging during the closing while driver was installed with `create_background_task = false` configuration.
Unit test `class_specific_requests_with_external_polling` has been added to the test cases.

### Problem
When the `create_background_task = false` configuration is used, the `event_handler_task` is not being created. Thus, the `hid_host_handle_events()` should be called periodically to handle the call `usb_host_client_handle_events()` outside the HID driver.
After `hid_host_install()` with `create_background_task = false` the `usb_host_client_deregister()` will never be called and the semaphore `all_events_handled` will remain taken. 

### Solution
Adding the `hid_host_all_events_was_handled()` to deregister the usb client and unlock the semaphore.

tore-espressif commented:
@roma-jam Thanks for the update, I think I understand the problem. I'd prefer if we could fix it without introducing new API. Could you please share your thoughts about this? :

1. Rename `s_hid_driver->end_client_event_handling` to `s_hid_driver->uninstall_request` (or something similar in meaning)
2. Extend `hid_host_handle_events()` with something like this:

``` c
esp_err_t hid_host_handle_events(uint32_t timeout)
{
    ...
    usb_host_client_handle_events(s_hid_driver->client_handle, timeout); // @todo add error checking
    if (s_hid_driver->uninstall_request) {
        usb_host_client_deregister(s_hid_driver->client_handle);
        xSemaphoreGive(s_hid_driver->all_events_handled); // @todo Is this really needed?
    }
    return ESP_OK;
}
```

The code snippet is pretty far from working, but the idea is clear: Could we handle the client deregistration in `hid_host_handle_events`? This way we do not need new API

roma-jam commented:
> Could we handle the client deregistration in `hid_host_handle_events`? This way we do not need new API

I thought about that, but I am not sure that there will be no collision with event handling and uninstall calling. 

Right now, all the events shoud be handled before uninstall, which is clear. As we are not able to handle the events after the uninstall completion. Which is also clear.

Moreover, there should be some api to set this `s_hid_driver->uninstall_request` and it cannot be done inside `hid_host_uninstall()`…

Otherwise, I am agreed with the idea of trying not to change api. 
I will think about it a little bit more.

roma-jam commented:
> There can be scenario where:
> 
> 1. `hid_host_install()` is called with `create_background_task = false`
> 2. User does clean up with `hid_host_uninstall()`
> 3. User does not call `hid_host_handle_events()` at all.
> 
> Who will give the semaphore then !!
> 
> `hid_host_uninstall()` will remain blocked on `xSemaphoreTake()`

Yes, indeed.
The point is that hid_host_uninstall() wait semaphore which handled only in background task. 
The next steps doesn't really matters, yes.

roma-jam commented:
Hey @tore-espressif, 

I have changed the logic and now the API is the same. 

> xSemaphoreGive(s_hid_driver->all_events_handled); // @todo Is this really needed?

Answering your question - Honestly, probably we do not need that. But in CDC Driver you do the same but with events handling (`CDC_ACM_TEARDOWN` and `CDC_ACM_TEARDOWN_COMPLETE`), so probably there is some idea behind that. 

@esp-saurabhbansal , 
Thanks for pointing that out. I have changed the logic during the `hid_host_uninstall()` function as well as have added the test case `class_specific_requests_with_external_polling_without_polling` when `hid_host_handle_events()` is not being called before uninstall the driver to verify the correct behavior in this scenario. 

Feel free to review changes, I am looking forward for yours feedback.
Thanks.

esp-saurabhbansal commented on usb/usb_host_hid/hid_host.c:
```diff
@@ -1254,13 +1277,29 @@ esp_err_t hid_host_device_close(hid_host_device_handle_t hid_dev_handle)
+
+        xSemaphoreGive(s_hid_driver->all_events_handled);
+        // wait for hid_host_uninstall completion
+        s_hid_driver->ext_task_config.external_task_hdl = xTaskGetCurrentTaskHandle();
```
`external_task_hdl` is set in function hid_host_handle_events() in one thread and then referred in function hid_host_uninstall() in another thread. These two threads can race against each other. I think, its better to set `external_task_hdl` at the very start if function hid_host_handle_events(). It will reduce the possibility race condition.

Something like this:

```
esp_err_t hid_host_uninstall(void)
{
    ***
    if (s_hid_driver->ext_task_config.active &&
        !s_hid_driver->ext_task_config.external_task_hdl) {
            ESP_LOGE(TAG, "HID Host external event handling has not been called");
            // Deregister client by ourselves and release semaphore
            HID_RETURN_ON_ERROR( usb_host_client_deregister(s_hid_driver->client_handle),
                                 "HID Client deregister failure");
            xSemaphoreGive(s_hid_driver->all_events_handled);
        }
    }

    xSemaphoreTake(s_hid_driver->all_events_handled, portMAX_DELAY);
    vSemaphoreDelete(s_hid_driver->all_events_handled);
    s_hid_driver->ext_task_config.external_task_hdl = NULL;
    s_hid_driver->ext_task_config.active = false;
    ***
}

esp_err_t hid_host_handle_events(uint32_t timeout)
{
    HID_RETURN_ON_FALSE(s_hid_driver,
                        ESP_ERR_INVALID_STATE,
                        "HID Driver is not installed");

    if (!s_hid_driver->ext_task_config.active) {
        return ESP_ERR_NOT_SUPPORTED;
    }
    s_hid_driver->ext_task_config.external_task_hdl = xTaskGetCurrentTaskHandle();

    esp_err_t ret = usb_host_client_handle_events(s_hid_driver->client_handle, timeout);

    if (s_hid_driver->end_client_event_handling) {
        HID_RETURN_ON_ERROR( usb_host_client_deregister(s_hid_driver->client_handle),
                             "HID Client deregister failure");

        xSemaphoreGive(s_hid_driver->all_events_handled);
    }

    return ret;
}
```

roma-jam replied:
Hey @esp-saurabhbansal, 
Thanks for the review.

Probably, I don't get the point. Let me try to explain my logic, probably after that you can just point the moment when I am not right or has a logic misleading. 

1. When application logic calls the `hid_host_handle_events()`, we spend some time in `usb_host_client_handle_events()`. To make it compatible with power management features such as tickless idle and automatic light sleep we need to provide some mechanism with a possibility to use maximum delay `portMAX_DELAY` to wait next event and do not poll peripheral every N times per minute. This can be achieved when we call `hid_host_handle_events(portMAX_DELAY);` from the external polling task;
2. In this case, the host will wait another client event. If there is no event, the logic will return from this function after `usb_host_client_unblock()` during the `hid_host_uninstall()`;
3. Before client unblock, we change the state of internal flag `s_hid_driver->end_client_event_handling = true;` during uninstall;
4. When logic returns from `usb_host_client_handle_events()` in `hid_host_handle_events()` fulfilled in external task,  the  flag end_client_event_handling is true and we unblock the client, release semaphore and save the Task handle and suspend the external Task (here is another question, if there is more logic except the client events polling in this task, so we can halt this logic as well, which is not so good);

...

I see. 
Do you mean, that we can Resume the external Task before it (external Task) release semaphore and will go to Suspend, right?

esp-saurabhbansal replied:
Lets call the thread as threadA which calls hid_host_uninstall() and threadB which calls hid_host_handle_events()

After Point 3, in case threadA keeps on executing and threadB does not get CPU, threadA will check for the if conditions and it finds external_task_hdl as NULL and it execute this block. It will call usb_host_client_deregister() -> xSemaphoreGive() -> ....
```
        if (NULL == s_hid_driver->ext_task_config.external_task_hdl) {
            ESP_LOGE(TAG, "HID Host external event handling has not been called");
            // Deregister client by ourselves and release semaphore
            HID_RETURN_ON_ERROR( usb_host_client_deregister(s_hid_driver->client_handle),
                                 "HID Client deregister failure");
            xSemaphoreGive(s_hid_driver->all_events_handled);
        }
```


Next, threadB gets the CPU and starts executing. After finding s_hid_driver->end_client_event_handling as true, it will start executing this block:
```
    if (s_hid_driver->end_client_event_handling) {
        HID_RETURN_ON_ERROR( usb_host_client_deregister(s_hid_driver->client_handle),
                             "HID Client deregister failure");

        xSemaphoreGive(s_hid_driver->all_events_handled);
        // wait for hid_host_uninstall completion
        s_hid_driver->ext_task_config.external_task_hdl = xTaskGetCurrentTaskHandle();
        vTaskSuspend(NULL); // Task will be resumed from hid_host_uninstall()
    }
```

This threaB will again call usb_host_client_deregister() -> xSemaphoreGive() -> sets external_task_hdl -> vTaskSuspend itself.

This flow happens because threadA find the value of external_task_hdl as NULL.
So, my sugession was to set external_task_hdl with xTaskGetCurrentTaskHandle(); before calling usb_host_client_handle_events().

I can be unclear now also. We can have a call to discuss.

esp-saurabhbansal reviewed (commented):
Code logic is very good.
Can you check my suggestion once, please.

Dazza0 commented on usb/usb_host_hid/hid_host.c:
```diff
@@ -377,6 +444,23 @@ static esp_err_t hid_host_add_interface(hid_device_t *hid_device,
+ * @return esp_err_t
+ */
+static esp_err_t _hid_host_remove_interface(hid_iface_t *hid_iface)
+{
+
+    hid_iface->state = HID_INTERFACE_STATE_NOT_INITIALIZED;
```
```suggestion
{
    hid_iface->state = HID_INTERFACE_STATE_NOT_INITIALIZED;
```

Dazza0 commented on usb/usb_host_hid/include/usb/hid_host.h:
```diff
@@ -16,6 +17,8 @@
 extern "C" {
 #endif
 
+#define HID_STR_DESC_SIZE           32
```
The max length of a string descriptor according to USB 2.0 specification is 256-2 bytes. Please add a description as to what this macro actually means.

```suggestion
#define HID_STR_DESC_MAX_SIZE           32
```

Dazza0 commented on usb/usb_host_hid/hid_host.c:
```diff
@@ -512,51 +596,53 @@ static esp_err_t hid_host_interface_shutdown(hid_host_device_handle_t hid_dev_ha
         HID_ENTER_CRITICAL();
-        iface->state = HID_INTERFACE_STATE_NOT_INITIALIZED;
-        STAILQ_REMOVE(&s_hid_driver->hid_ifaces_tailq, iface, hid_interface, tailq_entry);
-        free(iface);
+        _hid_host_remove_interface(interface);
```
There's a gap in the critical section here. Thus, during that gap, another thread could already have removed the interface from the `hid_ifaces_tailq`. What if we removed it from the list first and then carried out the shut down.

roma-jam replied:
Yes, I know that. And still I didn't come up with any better idea how to fix it. 
There is a logical problem with architecture, that is why I decided to change it some time ago. 
I explained everything more detailed in the document, which I attached to the PR for next version: https://github.com/espressif/idf-extra-components/pull/214#issuecomment-1643770574

The whole idea to keep list of usb device opened device is strange. 
That is why there is a two-way choice during the hid_host_unsinstall(). 

But I'll try to figure something out to fix this version. 
Thanks.

roma-jam replied:
Change the way how list is freed. 
Yes, there is still a gap in critical section, but in case if the some elements will be absent, should be no problem. 

I guess this is the best option to do in case of current logic.

Dazza0 reviewed (commented):
@roma-jam Seems like we have a pretty complex set of calls to remove devices/interfaces on `hid_host_uninstall()`. As I've mentioned previously, would it be easier if we just required all opened interfaces/devices to be closed before allowing users to call `hid_host_uninstall()`.

Having `hid_host_uninstall()` trying to remove live devices/interfaces could lead to some tricky concurrency issues,

tore-espressif commented on usb/usb_host_hid/include/usb/hid_host.h:
```diff
@@ -35,11 +48,27 @@ typedef enum {
+typedef struct {
+    wchar_t *p;
+    uint16_t len;
+} hid_host_dev_string_t;
```
Unused type definition?

roma-jam replied:
Yep, right. 
Thanks.

tore-espressif commented on usb/usb_host_hid/idf_component.yml:
```diff
@@ -1,4 +1,4 @@
-version: "1.0.0"
+version: "1.0.1~1"
```
1.0.1 should be enough, right?

roma-jam replied:
I think so.
But I have asked you about the description, because I understood that I don't really know when and what better to use. 

Anyway, changed to 1.0.1 right now.

tore-espressif replied:
No problem! Here are some details about versioning https://docs.espressif.com/projects/idf-component-manager/en/latest/reference/versioning.html#versioning-scheme

roma-jam replied:
Thanks, I have updated that and verifyed with esp-idf as well. 
New example works well with both versions: 1.0.0 and 1.0.1.

tore-espressif reviewed (approved):
Thank you for your patience @roma-jam !
I left few nitpicks, otherwise LGTM!

And please make sure that the HID example in esp-idf works OK with this new version!
